### ISSUE TITLE
Step6 Visualization results in automatic exit

### ISSUE THREAD

**Issue by Yfanmeng** [ktgiahieu/Fast-PPF#13]
Hi author, thank you very much for your open source project, I am running the code under Window system, during the run, everything is working fine in step 5, however, when I run to step 6, only the output is available, the visualization graph exits automatically and the program terminates. For your code, I did not make any changes. If you have time, could you help to check what went wrong, thank you very much. The experimental platform I am using is QT Creator and PCL library version 1.8.1.
Step 5
![1](https://github.com/ktgiahieu/Fast-PPF/assets/115934943/6550a6fa-65f8-4bff-9f23-8dbf9c2cb4a3)
Step 6     When I run to step 6, the visualization automatically exits and the code run terminates without running to step 7.
![2](https://github.com/ktgiahieu/Fast-PPF/assets/115934943/47a4db08-5776-4479-b5cf-99c3fb8b2b18)

**Comment by ktgiahieu** [comment]
Hi,

The problem was caused by the VTK library (used for visualization of this project). You might have installed the newest version of VTK, which might conflict with old libraries from this project.
It seems that this problem occured to this person as well: https://github.com/ktgiahieu/Fast-PPF/issues/6#issuecomment-1156064785

I would try his suggestion of downgrading VTK to 6.x and check if the problem persists.

Best,

**Comment by Yfanmeng** [comment]
Hi, 
I have solved it, but I used not the previous QT Creator experimental platform, replaced the experimental platform with VS2019 and PCL1.11, and finally I can realize the whole process, thanks again for your reply.

**Comment by Yfanmeng** [comment]
> Hi,
> 
> The problem was caused by the VTK library (used for visualization of this project). You might have installed the newest version of VTK, which might conflict with old libraries from this project. It seems that this problem occured to this person as well: [#6 (comment)](https://github.com/ktgiahieu/Fast-PPF/issues/6#issuecomment-1156064785)
> 
> I would try his suggestion of downgrading VTK to 6.x and check if the problem persists.
> 
> Best,


Dear Author.
Now I have another problem, I have implemented the whole process and can clearly see the matches, but the results are not very good, I saw your previous answer to someone else's question about needing to change the code for the PPF part of the PCL, which is answered by this picture later:
![1](https://github.com/ktgiahieu/Fast-PPF/assets/115934943/8c2d5947-0d52-43ce-9a01-7276413185fb)

Then I saw this quote from you
"In my day, I downloaded the PCL source code and rebuild the hold PCL library. BUT looking back, you can just add this new function in the code and replace pcl::computePPFPairFeature (line 357 of this file ) with this new function."
But I didn't understand this sentence very well, can you please tell me exactly in which file to add the new code and which line from line 357 needs to be replaced? Looking forward to your reply.

**Comment by ktgiahieu** [comment]
Hi,

Sorry I am away from my computer right now, so I can't help you with the exact details.

Basically, you need to define a new function (e.g customComputePairFeatures) with the implementation as in the image you sent me.

Then, you will replace the lines 357-362 of [this file](https://github.com/ktgiahieu/Fast-PPF/blob/master/include/PPF/MyPPFRegistration.hpp#L357) with the call from your customComputePairFeatures function.

**Comment by Yfanmeng** [comment]
Dear Author.


Really surprised！I didn't expect you to send me an email！！！


thank you very much for your reply, I will try the solution you mentioned. Wish you a wonderful day. Thanks again!





------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "ktgiahieu/Fast-PPF"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2023年10月11日(星期三) 下午5:14
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [ktgiahieu/Fast-PPF] Step6 Visualization results in automatic exit (Issue #13)





 
Hi,
 
Sorry I am away from my computer right now, so I can't help you with the exact details.
 
Basically, you need to define a new function (e.g customComputePairFeatures) with the implementation as in the image you sent me.
 
Then, you will replace the lines 357-362 of this file with the call from your customComputePairFeatures function.

**Comment by Yfanmeng** [comment]
> Hi,
> 
> Sorry I am away from my computer right now, so I can't help you with the exact details.
> 
> Basically, you need to define a new function (e.g customComputePairFeatures) with the implementation as in the image you sent me.
> 
> Then, you will replace the lines 357-362 of [this file](https://github.com/ktgiahieu/Fast-PPF/blob/master/include/PPF/MyPPFRegistration.hpp#L357) with the call from your customComputePairFeatures function.

Dear Author.
Following your method, I have solved the problem and the matching is indeed much better than before. Thanks again.

**Comment by Yfanmeng** [comment]
> Hi,
> 
> Sorry I am away from my computer right now, so I can't help you with the exact details.
> 
> Basically, you need to define a new function (e.g customComputePairFeatures) with the implementation as in the image you sent me.
> 
> Then, you will replace the lines 357-362 of [this file](https://github.com/ktgiahieu/Fast-PPF/blob/master/include/PPF/MyPPFRegistration.hpp#L357) with the call from your customComputePairFeatures function.

Dear Author.
Once again, I am disturbing you, I am experiencing the same problem that someone has encountered before, as shown in the picture below: 
![1](https://github.com/ktgiahieu/Fast-PPF/assets/115934943/eb02f707-867f-4d87-adf6-91e1e126c2df)

As you can see in this image, the top and rightmost round parts are not matched, and I think the top and rightmost round parts are easily matched. Then you gave the solution below:
![2](https://github.com/ktgiahieu/Fast-PPF/assets/115934943/c26e81bc-6bfb-4c91-baf1-69d68242fca5)

Of the three solutions you gave, I tried lowering the parameters in the first and second, but the results are still not very good. The first picture is the result of the modification according to the third solution, I changed the parameter double t_sampling = 0.04 to 0.02 to get the result.What other parameters should I adjust if I want to get better recognition? Looking forward to your reply.

**Comment by ktgiahieu** [comment]
Hi,

The method and my cheap 3D camera is not perfect (as you can see there are a lot of black blobs on the surface of the objects on the top, because of the reflection of the surface), so for the final implementation, I added 1 more condition for Pose Verification. As you can see in the attached video, there is a blue pose (the selected pose to pick up) among green poses. When performing Pose Verification, I decreased the thresholds to took into account more poses (include more false poses) and I also took into account the z value of the pose:
- With the lower z (closer to the camera), there will be less points captured by the camera (black blobs as mentioned earlier because of the reflection), hence the threshold to filter the poses should be lower
- With the higher z (further from the camera), there will be more points captured by the camera, hence the threshold to filter the poses should be higher.

Please note that with an industrial 3D camera with high quality, the steps above are unnecessary.
